### DISCUSSION TITLE
retrieving tokens from TokenHolder interface

### DISCUSSION THREAD

**Discussion by itachiuchihamadara** [OWASP/www-project-csrfguard#98] [Q&A] [answered]
Do you have any idea on how to retrieve tokens that are stored in InMemoryTokenHolder? In our project with version 3.1.0 we usually compare CSRF tokens from request with httpsession [since HttpSession contains csrf tokens as an attribute] and check equality. However, from version 4.0.0 we cannot retrieve csrf token attribute from sessions. In this case what can be done and how can I retrieve tokens from TokenHolder to be compared with request token since httpsession no longer contain csrf attribute?

**Comment by forgedhallpass** [comment, marked as answer]
Hello @itachiuchihamadara 

I am not sure I understand why would you want to access the CSRF tokens. Everything should happen seamlessly within CSRF Guard.

**Reply by forgedhallpass** [reply]
Something is not right on your side. There is no `validateCsrfToken` method in the latest version of CSRF Guard, also the `doFilter` line number doesn't match either. See: https://github.com/OWASP/www-project-csrfguard/blob/74008a23389bfd40131446e4bda794a0f925fe3a/csrfguard/src/main/java/org/owasp/csrfguard/CsrfGuardFilter.java#L58

As mentioned earlier, the CSRF token validation happens seamlessly. Please make sure to use the right dependencies: https://github.com/OWASP/www-project-csrfguard#using-with-maven

The same goes for the JavaScript and properties files as well.

To see a working setup where you can easily debug or play around, please refer to the provided test application:
https://github.com/OWASP/www-project-csrfguard/tree/master/csrfguard-test/csrfguard-test-jsp

**Reply by forgedhallpass** [reply]
I still don't understand your use-case. Request tokens are validated in the code I've linked above. In version `4.x`, CSRFGuard doesn't store CSRF tokens on the session anymore, but through the `TokenHolder` interface. This abstraction was introduced to also support stateless(sessionless) applications. The `InMemoryTokenHolder` is the default implementation, but you can also create your own storage (e.g. database) if you want by implementing the interface and setting the `org.owasp.csrfguard.TokenHolder` in your `csrfguard.properties` file (see https://github.com/OWASP/www-project-csrfguard/blob/74008a23389bfd40131446e4bda794a0f925fe3a/csrfguard/src/main/resources/csrfguard.properties#L447)

Furthermore, CSRF attacks are limited to a browser context (which is handled by CSRFGuard already), hence for machine-to-machine(m2m) interactions it is not needed.

Have you checked out the test application like I've recommended?
